flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSource.java

```diff
@@ -111,24 +112,17 @@
 		return kafkaSource;
 	}
 
-	@Override
-	public int getNumberOfFields() {
-		return fieldNames.length;
-	}
-
-	@Override
 	public String[] getFieldsNames() {
-		return fieldNames;
+		return TableSource$class.getFieldsNames(this);
```

Comment:
Hi,

I'm not sure about implementing this as a Scala trait with implemented methods. IMO, this makes it much harder to implement TableSources in Java (esp. for users who are not familiar with Scala and its implications). 

What do you think about implementing `TableSource` as abstract class and providing three other abstract classes that extend `TableSource` with the batch, the stream, and both interfaces?